props/saber_BC_buttons.h

```diff
@@ -386,31 +1136,297 @@ EFFECT(volmax);     // for maximum volume reached
+    if (current_volume_ < max_volume_) {
+      current_volume_ += max_volume_ * 0.10;
+    }
+    dynamic_mixer.set_volume(current_volume_);
+    mode::getSL<SPEC>()->SayVolumeUp();
+  }
+
+  void prev() override {
+    int current_volume_ = dynamic_mixer.get_volume();
+    if (current_volume_ > min_volume_) {
+      current_volume_ -= max_volume_ * 0.10;
+    }
+    dynamic_mixer.set_volume(current_volume_);
+    mode::getSL<SPEC>()->SayVolumeDown();
+  }
+
+  void QuickMaxVolume() {
+    dynamic_mixer.set_volume(max_volume_);
+    PVLOG_NORMAL << "** Maximum Volume\n";
+    mode::getSL<SPEC>()->SayMaximumVolume();
+  }
+
+  void QuickMinVolume() {
+    dynamic_mixer.set_volume(min_volume_);
+    PVLOG_NORMAL << "** Minimum Volume\n";
+    mode::getSL<SPEC>()->SayMinimumVolume();
+  }
+
+  void update() override {
+    SaberBase::DoEffect(EFFECT_VOLUME_LEVEL, 0);
+    // Overridden to substitute the tick sound
+    this->say_time_ = Cyclint<uint32_t>(millis()) + 1;
+    if (!this->say_time_) this->say_time_ += 1;
+  }
+
+  void select() override {
+    PVLOG_NORMAL << "** Saved - Exit Volume Menu\n";
+    mode::getSL<SPEC>()->SaySave();
+    SPEC::SteppedMode::select();
+  }
+
+  void exit() override {
+    PVLOG_NORMAL << "** Cancelled - Exit Volume Menu\n";
+    percentage_ = initial_percentage_;
+    dynamic_mixer.set_volume(initial_volume_);
+    mode::getSL<SPEC>()->SayCancel();
+    SPEC::SteppedMode::exit();
+  }
+
+  bool mode_Event2(enum BUTTON button, EVENT event, uint32_t modifiers) override {
+    switch (EVENTID(button, event, 0)) {
+      // Custom button controls for BCVolumeMode
+      case EVENTID(BUTTON_POWER, EVENT_FIRST_HELD_MEDIUM, 0):
+        QuickMaxVolume();
+        return true;
+
+      case EVENTID(BUTTON_POWER, EVENT_SECOND_HELD_MEDIUM, 0):
+      case EVENTID(BUTTON_AUX, EVENT_FIRST_HELD_MEDIUM, 0):
+        QuickMinVolume();
+        return true;
     }
+    // Use the select and exit controls from SelectCancelMode
+    return SPEC::SelectCancelMode::mode_Event2(button, event, modifiers);
+  }
+};
+
+#ifdef DYNAMIC_BLADE_LENGTH
+int selected_blade_;
+
+#ifdef BC_DUAL_BLADES
+template<class SPEC>
+struct BCSelectBladeMode : public SPEC::MenuBase {
+  int current_blade() { return this->pos_ + 1; }
+  ShowColorSingleBladeTemplate<Pulsing<ShowColorStyle,Black,800>,Pulsing<ShowColorStyle,Black,800>> highlighted_blade_;
+
+  uint16_t size() override {
+    return NUM_BLADES;
+  }
+  void mode_activate(bool onreturn) override {
+    mode::getSL<SPEC>()->SayEditBladeLength();
+    mode::getSL<SPEC>()->SaySelectBlade();
+    ShowColorStyle::SetColor(Color16(65535, 65535, 65535));
+    SPEC::SteppedMode::mode_activate(onreturn);
+    highlighted_blade_.Start(current_blade());
+    PVLOG_NORMAL << "** Highlighting blade: " << current_blade() << "\n";
+    say();
+  }
+
+  void mode_deactivate() {
+    highlighted_blade_.Stop(current_blade());
+  }
+
+  void next() override {
+    highlighted_blade_.Stop(current_blade());
```

Comment:
An alternate way to do this (which may or may not be better, I'm not actually sure) would be to save which blade is highlighted, and then do something like:

if (current_blade() != highlighted_blade_num_) {
  highlighted_blade_.Stop(highlighted_blade_num_);
  highlighted_blade_num_ = current_blade();
  highlighted_blade_.Stop(highlighted_blade_num_);
}
 
This code could be in update(), which means you would only need to do it once.
(Note the current code is fine, this is an optional idea.)

Reply:
ok to leave for now? Maybe if a non-prop version develops in the future and uses that way, then the prop version could adjust to follow suit?